package.json

```diff
@@ -52,7 +52,10 @@
     "betterer": "betterer",
     "betterer:merge": "betterer merge",
     "betterer:stats": "ts-node --transpile-only --project ./scripts/cli/tsconfig.json ./scripts/cli/reportBettererStats.ts",
-    "betterer:issues": "ts-node --transpile-only --project ./scripts/cli/tsconfig.json ./scripts/cli/generateBettererIssues.ts"
+    "betterer:issues": "ts-node --transpile-only --project ./scripts/cli/tsconfig.json ./scripts/cli/generateBettererIssues.ts",
+    "build:plugin": "TS_NODE_PROJECT=\"${0}/tsconfig.json\" webpack -c ./tools/plugins/shared/webpack/webpack.config.ts --env production --env dir=${0}",
```

Comment:
I was expecting this to call a build command in the grafana-testdatasouce package.json but there doesn't appear to be one. e.g. `yarn workspaces @grafana-plugins/testdatasource run build` or `lerna run build --scope=@grafana-plugins`. 

nit: I would name this `plugin:build` to differentiate it from the `build`, `build:nominify` scripts that build the grafana application and align it with the `packages:<cmd>` scripts format.

Reply:
yeah, that's the key point of this proposal, not having a package.json per plugion (therefore, not having more packages). The benefits are: less maintenance and less duplicated code. The con is more coupling with core-grafana since we are forcing to use the same deps (which I guess we do anyway for sub packages).

This is the minimal incremental step, if in the future we think we want to add packages for plugins we can do so.

re: naming nit :+1: will change it.

Reply:
I'm all for incremental steps but IMHO making them workspaces feels unavoidable as more and more plugins are extracted.

With this setup how does a plugin extend the webpack config in `tools/plugins/shared/webpack`?

The dependency coupling with core will be problematic if an extracted plugin wants to use a different version of a 3rd party package. We're also mixing more dependencies into the root package.json which I believe makes it more difficult to maintain. Another downside is without workspaces we lose all task orchestration and parallel execution that yarn workspaces / lerna can provide. We lose `yarn workspace focus` which allows installation of only that workspaces dependencies which can be useful for things like CI.

I guess we can re-evaluate it if / when it becomes an issue. 👍

Reply:
yeah, I agree with the downsides of this but the alternative also has some. Some of the push back in https://github.com/grafana/grafana/pull/72205 (which uses workspaces) is that having to maintain +50 workskpaces / packages.json + the boilerplate files (.eslintrc, jest config, etc) would be a burden that could be bigger.

This seems to be the safest decision, which is continuing as much as possible with how things work today, more like a monorepo approach.

In any case, as you say, this can be easily converted to a workspace in the future if we need it.

Reply:
I just familiarised myself with the other (https://github.com/grafana/grafana/pull/72205) PR, so it can be that I am missing something, in that case sorry.

I kind of share the same concerns Jack has raised, namely that plugins will not have the freedom to customise their dependencies or their config. Regarding the maintenance, couldn't they just extend (inherit) the shared configs, and then only customise them if they need to? I know it would mean more files and more boilerplate, but I don't think we would need to maintain them on a regular basis - until they are customised.

Although this is an incremental change, I am wondering what happens if at some point a plugin decides to customise its build config and its dependencies - how is that going to work with the common `build:plugin` command?

Reply:
I'm struggling to find the push back related to using workspaces in the PR you've linked to. 🤔 

In that previous PR the workspace config files could have been shared out to a shared location (as you're doing in this PR) and the datasource configs cleaned up so it would only need duplicated `tsconfig.json` and `webpack.config.ts` to extend from the shared configs. That feels like we'd be taking advantage of workspaces, allow for webpack config sharing, and isolated dependencies per plugin without so much duplication.

My biggest worry is that issues don't start to surface here until we're 10 or 15 plugins in and then we need to rework the whole setup. Maybe I'm thinking too far ahead.

Reply:
I honestly don't have a strong opinion here, I get both points so I am good with what the majority decides. Using `renovate` for automatically updating dependencies, having multiple packages does not seem a burden and give plugin devs that flexibility.

> I'm struggling to find the push back related to using workspaces in the PR you've linked to. 🤔

Right, that was mostly offline converstations. Let me add @torkelo to the loop since he was vocal in this. Please take a look!

Reply:
It's tricky, I don't know what is best, was just afraid of the explosion of config files of all different sorts for every plugin. Add to that having to maintain 10x package json files with all different dependencies. Each plugin in kibana looks so much slimmer in comparison: https://github.com/elastic/kibana/tree/main/x-pack/plugins/asset_manager   (but they bundle and version everything together, so maybe not a fair comparison). 

![image](https://github.com/grafana/grafana/assets/10999/788e1fe3-d4ea-499f-a239-a98d564f6971)

Maybe there are ways to minimize the amount of config files? 

It would be interesting to explore the plugins-bundled directory (it's a workspace) see if we can make this concept work for core data sources? See if we make the input data source there a backend data source, compiled into grafana binary, but the frontend is built using a separate webpack command.  

I think having separate webpacks builds for core data sources is not so bad? (I mean they would only be built this way, and not via https://github.com/grafana/grafana/blob/main/public/app/features/plugins/built_in_plugins.ts#L2  

wonder why https://github.com/grafana/grafana/blob/main/plugins-bundled/internal/input-datasource/package.json is so slim compared to https://github.com/grafana/grafana/blob/36b7d91adf862eb719982148cf6a2c9875b34b10/public/plugins/grafana-testdata-datasource/package.json ?

Reply:
Maybe one webpack with an entry point for ever "bundled" plugin could work?

Reply:
> Maybe there are ways to minimize the amount of config files?

we can definetly remove some files from there and still use workspaces (that is just what `create-plugin` generates by default).

> It would be interesting to explore the plugins-bundled directory (it's a workspace) see if we can make this concept work for core data sources? 

Yes, we can make this a workspace :+1: 

> I mean they would only be built this way, and not via https://github.com/grafana/grafana/blob/main/public/app/features/plugins/built_in_plugins.ts#L2

I think that, at least for the moment, we will keep the "built_in" way of building the plugin when it's running in "core" mode. The alternative would mean that core plugins would need to be built to be able to use them. This is an important change that involves a difference in the way Grafana devs work so it deserves a different conversation. For the moment, the "core" experience won't change in anything.

> wonder why https://github.com/grafana/grafana/blob/main/plugins-bundled/internal/input-datasource/package.json is so slim compared to https://github.com/grafana/grafana/blob/36b7d91adf862eb719982148cf6a2c9875b34b10/public/plugins/grafana-testdata-datasource/package.json 

I don't think the `input-datasource` can be considered a real-world plugin so it only has some minimal dependencies. The current `package.json` that the `create-plugin` is more complete but it's mostly devDependencies so nothing that affect runtime. In any case we can try to create one shared package that defines common dependencies and plugins would need to only define custom dependencies.

> Maybe one webpack with an entry point for ever "bundled" plugin could work?

I don't think that's what we want. We would lose the ability to build a single plugin and also the possibility to release plugins independently.

Reply:
> I don't think that's what we want. We would lose the ability to build a single plugin and also the possibility to release plugins independently.
Not 100% sure that would be the case. Just because we could build many plugins at once they would still have independent entry points/outputs. But maybe this is not feasible with webpack. 

>I think that, at least for the moment, we will keep the "built_in" way of building the plugin when it's running in "core" mode. The alternative would mean that core plugins would need to be built to be able to use them.

Hm.. think this requirement is adding some unnecessary complexity here. Having to build plugins using a separate command is not a huge problem (we have been talking about the need to split up the single large webpack build for some time). I think it would be a problem if each individual plugin needed a separate command and each came with a slow startup, so building all would be very slow, hence the idea to have a single webpack that builds all "standalone but bundled" plugins. 

From my point of view de-coupling data sources have two challenges
* Not needing separate binaries in the main distribution (so we can compile in the bundled ds backends) 
* Not requiring separate repos for each, so we can group them together, and build and version them together for a faster dev experience. 

having the "bundled" plugins require a separate build step for the frontend assets I think is ok? 

On very different angle, maybe splitting data source plugins into two is a good idea? Backend and frontend as separate "plugins" / updatable things. Sounds like it would increase complexity in deployment configurations but if we look ahead, with multi-tenant "remote" ds backends, not having to update and deploy the backend for a small UI fix in a query editor sounds useful but maybe they don't need to be separate plugins for that to be possible.

Reply:
> I think it would be a problem if each individual plugin needed a separate command and each came with a slow startup, so building all would be very slow, hence the idea to have a single webpack that builds all "standalone but bundled" plugins.

That defeats the idea of having independent plugins. We may decouple them from grafana but they would still be coupled to each other.

> * Not needing separate binaries in the main distribution (so we can compile in the bundled ds backends)
> * Not requiring separate repos for each, so we can group them together, and build and version them together for a faster dev experience.

That is as Grafana is today, that won't change

> having the "bundled" plugins require a separate build step for the frontend assets I think is ok?

IMO, to not diverge of the current core experience I wouldn't force people to run extra commands if plugins are meant to run in "core" mode (as of today).

> On very different angle, maybe splitting data source plugins into two is a good idea? Backend and frontend as separate "plugins" / updatable things. Sounds like it would increase complexity in deployment configurations but if we look ahead, with multi-tenant "remote" ds backends, not having to update and deploy the backend for a small UI fix in a query editor sounds useful but maybe they don't need to be separate plugins for that to be possible.

That may be interesting to explore but out of the scope for the moment.

Reply:
> IMO, to not diverge of the current core experience I wouldn't force people to run extra commands if plugins are meant to run in "core" mode (as of today).

Why not? As long as it fast, and not 10 extra commands. (And that there would be a single command that can do everything that is needed build the different parts of the frontend). 

Frontend data source plugins have separate webpack chunks, and loaded async today so that part would not change (in terms of performance). The main things we care about is bundling (shipping a single package that is as small as possible) and as easy to maintain and develop as possible. 

> That defeats the idea of having independent plugins. We may decouple them from grafana but they would still be coupled to each other.

True they would be coupled to a single build webpack build, but asset wise they could be independent. And it feels like it would be easy to modify that webpack command to only build a single plugin. But yes, dependencies might be coupled this way, not sure how flexible webpack is here. Our packages/** have different dependencies but are build with out main / single webpack?

Reply:
>I think it would be a problem if each individual plugin needed a separate command and each came with a slow startup, so building all would be very slow, hence the idea to have a single webpack that builds all "standalone but bundled" plugins.

I think task orchestration could be handled by lerna so we provide a single command to build all plugins or optionally make use of `--scope` to limit it to only build the plugins we care about. (e.g. `yarn plugins:build --scope="@grafana-plugins/grafana-testdata-datasource"`). This can also be the same for tasks like testing and linting. Isolating these commands to only run on the code that is being modified during development and CI should give plugin devs less friction and faster turn arounds.

Adding in flags to parallelise builds to make use of additional cores (given none of these plugins should depend on each other) will likely speed things up too. These webpack configs can cache builds locally too for an additional speed boost.

I'm not sure how much of a difference (speed wise) there will be between throwing all plugins at a single webpack config vs spinning up n number of processes concurrently. Feels like the downside here however we approach it would be slow builds when there are a lot of plugins that need to be built. 

>But yes, dependencies might be coupled this way, not sure how flexible webpack is here.

The dependencies are defined in the package.json. To the best of my knowledge Webpack doesn't know any more than to follow the import statements. Unless each plugin is its own workspace the only way I think a plugin can use its own version of an NPM package is aliasing the package in package.json (ex. `"lodash5": npm:lodash@5.0.0` then in the ts file `import {debounce} from 'lodash5'`).

Reply:
Looking better! 

> Re: Remove the plugin from built_in_plugins to use its module.js 

yea, think that makes a ton of sense and then we can move it out of public and into the bundled plugins directory, and just feels like it makes it so much simpler to have the plugin built only one way (at least the frontend).